api/progress.go

```diff
@@ -71,33 +74,86 @@ type ProgressRequest struct {
 	// Note: To be able to save the progress, we also need the userID, but it`s already contained in the Gin context
 }
 
+// saveProgress saves progress to a buffer that is flushed at a fixed interval.
 func saveProgress(c *gin.Context) {
 	var request ProgressRequest
-
 	err := c.BindJSON(&request)
 
 	if err != nil {
-		log.WithError(err).Warn("Could not bind JSON from progressReport.")
+		log.WithError(err).Warn("Could not bind JSON.")
 		c.AbortWithStatus(http.StatusBadRequest)
 		return
 	}
-
 	foundContext, exists := c.Get("TUMLiveContext")
-
 	if !exists {
 		return
 	}
-
 	tumLiveContext := foundContext.(tools.TUMLiveContext)
-
 	if tumLiveContext.User == nil {
 		c.AbortWithStatus(http.StatusForbidden)
 		return
 	}
+	streams, err := dao.GetStreamsByIds([]uint{request.StreamID})
+
+	if err != nil {
+		log.WithError(err).Warn("Could not get stream.")
+		c.AbortWithStatus(http.StatusInternalServerError)
+		return
+	}
 
 	progressBuff.add(model.StreamProgress{
 		Progress: request.Progress,
 		StreamID: request.StreamID,
 		UserID:   tumLiveContext.User.ID,
+		CourseID: streams[0].CourseID,
 	})
 }
+
+// WatchedRequest corresponds the request that is sent when a user marked the video as watched on the watch page.
+type WatchedRequest struct {
+	StreamID uint `json:"streamID"`
+	Watched  bool `json:"watched"`
+}
+
+// markWatched marks a VoD as watched in the database.
+func markWatched(c *gin.Context) {
+	var request WatchedRequest
+	err := c.BindJSON(&request)
+	if err != nil {
+		log.WithError(err).Warn("Could not bind JSON.")
+		c.AbortWithStatus(http.StatusBadRequest)
+		return
+	}
+	foundContext, exists := c.Get("TUMLiveContext")
+	if !exists {
+		return
+	}
+	tumLiveContext := foundContext.(tools.TUMLiveContext)
+	if tumLiveContext.User == nil {
+		c.AbortWithStatus(http.StatusForbidden)
+		return
+	}
+	streams, err := dao.GetStreamsByIds([]uint{request.StreamID})
```

Comment:
Why not `getStreamByID`?

Reply:
Good point, I forgot to change it after refactoring.